src/Framework/Framework/Compilation/ViewCompiler/DefaultViewCompilerCodeEmitter.cs

```diff
@@ -0,0 +1,360 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using DotVVM.Framework.Binding;
+using System.Collections.Concurrent;
+using System.Linq.Expressions;
+using Microsoft.Extensions.DependencyInjection;
+using DotVVM.Framework.Controls;
+
+namespace DotVVM.Framework.Compilation
+{
+    public class DefaultViewCompilerCodeEmitter
+    {
+        private static Type[] emptyTypeArguments = new Type[] { };
+
+        private int CurrentControlIndex;
+
+        public const string ControlBuilderFactoryParameterName = "controlBuilderFactory";
+        public const string ServiceProviderParameterName = "services";
```

Comment:
```suggestion
        public readonly ParameterExpression ServiceProviderParameter = Expression.Parameter("services", typeof(IServiceProvider));
```

With Linq Expressions, you can do this ;) (and the same for the ControlBuilderFactoryParameterName). The same expression can be reused in multiple runs without problems and no need to lookup it by string in some dictionary.

Reply:
I dont like this solution very much as "services" it is passed as an argument to the block, and therefore is added on block creation. This provides certain degree compile time checking. Also I find it more systematic.